File: src/main/java/at/yawk/math/algorithm/LcmSolver.java
package at.yawk.math.algorithm;

import at.yawk.math.data.Expressions;
import at.yawk.math.data.IntegerExpression;
import java.math.BigInteger;
import lombok.experimental.UtilityClass;

/**
 * @author yawkat
 */
@UtilityClass
public class LcmSolver {
    public static IntegerExpression lcm(IntegerExpression a, IntegerExpression b) {
        if (!a.isPositive()) { throw new IllegalArgumentException(a + " is not positive"); }
        if (!b.isPositive()) { throw new IllegalArgumentException(b + " is not positive"); }
        BigInteger x = a.getValue();
        BigInteger y = b.getValue();
        return Expressions.integer(x.multiply(y).divide(x.gcd(y)));
    }
}

File: src/main/kotlin/at/yawk/math/algorithm/LcmSolver.kt
package at.yawk.math.algorithm

import at.yawk.math.data.Expression
import at.yawk.math.data.Expressions
import at.yawk.math.data.IntegerExpression
import at.yawk.math.data.Sign

/**
 * @author yawkat
 */
object LcmSolver {
    fun lcm(a: IntegerExpression, b: IntegerExpression): Expression {
        if (a.sign != Sign.POSITIVE) {
            throw IllegalArgumentException("$a is not positive")
        }
        if (b.sign != Sign.POSITIVE) {
            throw IllegalArgumentException("$b is not positive")
        }
        return Expressions.int(a.value.multiply(b.value).divide(a.value.gcd(b.value)))
    }
}